Add Next Run to UI

bbovenzi opened the pull request:
Adds info on the next run from `next_dagrun` in the hoem page list of all dags and on the individual dag pages.

#17726

Unfortunately, `next_dagrun` only existed on `DagModel` and not `Dag` so I had to manually add it to every relevant endpoint in `views.py`. I also had to do some weird js logic to format the dates in line with the rest of the UI.

~~Also, I tried to update the copy on execution date. Open to suggestions.~~ (Updating use of `execution_date` throughout the UI will be a different PR)

<img width="1619" alt="Screen Shot 2021-08-19 at 4 15 18 PM" src="https://user-images.githubusercontent.com/4600967/130084389-8bf90487-d04a-440e-a8db-c73a50237fa1.png">

<img width="1018" alt="Screen Shot 2021-08-19 at 4 14 59 PM" src="https://user-images.githubusercontent.com/4600967/130084426-e8db0bb6-9733-43a5-992f-07e29bb93143.png">

Closes #2460

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

bbovenzi committed:
add next run to home page

bbovenzi committed:
add nextrun to dag pages

bbovenzi committed:
keep date check consistent

bbovenzi committed:
fix test_views

ashb commented on airflow/www/templates/airflow/dags.html:
```diff
@@ -98,7 +98,10 @@ <h2>{{ page_title }}</h2>
+                <span class="material-icons text-muted js-tooltip" aria-hidden="true" title="Logical Date/Time of the latest Dag Run.">info</span>
+              </th>
+              <th style="width:180px;">Next Run
+                <span class="material-icons text-muted js-tooltip" aria-hidden="true" title="Expected Date/Time of the next Dag Run.">info</span>
```
What date does this show? Is it the logical date that is going to be created, or the time _when_ the run will be created?

I think we need to display both of these (though one should be in a tooltip I think)

uranusjr replied:
Good point, it’s the logical date I think. The next expected *run* time is available in `DagModel.next_dagrun_create_after`.

This raises another issue though—perhaps we should ditch all the logical date labels, and always display the current/next DAG run’s run date and data interval instead.

ashb replied:
@uranusjr Hmmm that might remove the ambiguity, yeah

Do we retroactively set data_interval on old runs when in our migration?

uranusjr replied:
We don’t, but those can be calculated on the fly rather cheaply when needed. Or we could just display `data_interval_start` as *Unknown* (perhaps with a tooltip explaining why).

uranusjr replied:
Or something like

```jinja
{% if dagrun.data_interval_start is not None or dagrun.data_interval_end is not None %}
    {# Render data interval #}
{% else %}
    {# Render logical date #}
{% endif %}
```

Since users probably would want to see `execution_date` for those old runs?

bbovenzi replied:
Right now we get the last run execution date and start time here: https://github.com/apache/airflow/blob/main/airflow/www/views.py#L919

How should we render it with data_interval?

uranusjr replied:
Oh… that’s a bit complicated due to how `query` is constructed… I’ll add a commit to your branch for this later (need to do something first).

uranusjr replied:
Pushed, I think this should work…

bbovenzi replied:
Actually, we use `execution_date` in a lot more parts of the UI. It should be a separate PR.

ashb commented on airflow/www/templates/airflow/dag.html:
```diff
@@ -84,8 +84,17 @@ <h4 class="pull-right" style="user-select: none;-moz-user-select: auto;">
       </a>
+      {% if dag_model is defined and dag_model.next_dagrun is defined %}
+        <p class="label label-default" style="margin-left: 5px">
+          Next Run: <span id="next-run">{{ dag_model.next_dagrun }}</span>
```
Make this a `<time>` element maybe?

bbovenzi replied:
`<time>` doesn't really help us here

ryanahamilton commented on airflow/www/templates/airflow/dags.html:
```diff
@@ -170,12 +173,16 @@ <h2>{{ page_title }}</h2>
                   <svg height="10" width="10" id='task-run-{{ dag.safe_dag_id }}' style="display: block;"></svg>
                 </td>
-                {# Column 8: Actions #}
+                {# Column 9: Actions #}
```
We should probably just remove all of these comments. They're a bit superfluous and more nuisance than helpful IMO.

ryanahamilton commented on airflow/www/templates/airflow/dags.html:
```diff
@@ -170,12 +173,16 @@ <h2>{{ page_title }}</h2>
                 </td>
-                {# Column 7: Recent Tasks #}
+                {# Column 7: Next Run #}
+                <td class="text-nowrap next-run">
```
We've been using a `js-` prefix for classes used for scripting (not styling) purposes.
```suggestion
                <td class="text-nowrap js-next-run">
```

uranusjr committed:
Include data interval values in /last_dagruns view

bbovenzi committed:
Use dag.next_dagrun_create_after

bbovenzi committed:
merge main

ashb commented on airflow/www/templates/airflow/dag.html:
```diff
@@ -84,8 +84,17 @@ <h4 class="pull-right" style="user-select: none;-moz-user-select: auto;">
       </a>
+      {% if dag_model is defined and dag_model.next_dagrun_create_after is defined %}
+        <p class="label label-default" style="margin-left: 5px">
+          Next Run: <span id="next-run">{{ dag_model.next_dagrun_create_after }}</span>
```
By making this a `<time>` element it will automatically get converted by the [timezone selector](https://github.com/apache/airflow/blob/e4d3da16bad66aa40549cedac325d02e6b6559a9/airflow/www/static/js/datetime_utils.js#L74) (not to mention that just semantically this _is_ a time :grin:)

bbovenzi replied:
Oh yes, I forgot that we select elements by `<time>` in `datetime_utils.js`. Fixed.

bbovenzi committed:
Fix timezone formatting

Use `<time>` to use our existing `datetime_utils` to format and handle timezone changes

bbovenzi committed:
Update next and last run tooltips

uranusjr reviewed (approved):
Don’t really understand the js code, but the end result looks awesome

uranusjr committed:
Merge branch 'main' into add-next-run

bbovenzi committed:
Merge branch 'main' into add-next-run

bbovenzi committed:
wrap meta tags in if statement
